libyara/modules/dex.c

```diff
@@ -0,0 +1,650 @@
+    uint32_t access_flags = *class_defs[i].access_flags;
+    uint32_t superclass_idx = *class_defs[i].superclass_idx;
+    uint32_t interfaces_offset = *class_defs[i].interfaces_offset;
+    uint32_t source_file_idx = *class_defs[i].source_file_idx;
+    uint32_t annotations_offset = *class_defs[i].annotations_offset;
+    uint32_t class_data_offset = *class_defs[i].class_data_offset;
+    uint32_t static_values_offset = *class_defs[i].static_values_offset;
+
+    set_integer(class_idx, module, "class_defs[%i].class_idx", i);
+    set_integer(access_flags, module, "class_defs[%i].access_flags", i);
+    set_integer(superclass_idx, module, "class_defs[%i].superclass_idx", i);
+    set_integer(interfaces_offset, module, "class_defs[%i].interfaces_offset", i);
+    set_integer(source_file_idx, module, "class_defs[%i].source_file_idx", i);
+    set_integer(annotations_offset, module, "class_defs[%i].annotations_offset", i);
+    set_integer(class_data_offset, module, "class_defs[%i].class_data_offset", i);
+    set_integer(static_values_offset, module, "class_defs[%i].static_values_offset", i);
+  }
+  yr_free(class_defs);
+}
+
+
+void load_proto_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = *dex_header->proto_ids_offset;
+  int proto_ids_size = sizeof(PROTO_ID_ITEM[*dex_header->proto_ids_size]);
+  if (offset + proto_ids_size > data_size) {
+    set_integer(1, module, "invalid_dex");
+    return;
+  }
+
+  PROTO_ID_ITEM *proto_ids = yr_malloc(proto_ids_size);
+  memcpy(proto_ids, data + offset, proto_ids_size);
+
+  proto_ids_size = *dex_header->proto_ids_size * sizeof(PROTO_ID_ITEM);
+  int i,p;
+  for (i = 0, p = 0; p < proto_ids_size; i += 1, p += sizeof(PROTO_ID_ITEM)) {
+    uint32_t shorty_idx = *proto_ids[i].shorty_idx;
+    uint32_t return_type_idx = *proto_ids[i].return_type_idx;
+    uint32_t parameters_offset =  *proto_ids[i].parameters_offset;
+
+    set_integer(shorty_idx, module, "proto_ids[%i].shorty_idx", i);
+    set_integer(return_type_idx, module, "proto_ids[%i].return_type_idx", i);
+    set_integer(parameters_offset, module, "proto_ids[%i].parameters_offset", i);
+
+    //printf("shorty_idx(%i) = %s\n", shorty_idx, get_string_item(shorty_idx, module));
+  }
+  yr_free(proto_ids);
+}
+
+
+void load_field_ids(PDEX_HEADER dex_header, uint8_t *data, size_t data_size, YR_OBJECT *module) {
+  uint32_t offset = *dex_header->field_ids_offset;
+  int field_ids_size = sizeof(FIELD_ID_ITEM[*dex_header->field_ids_size]);
+  if (offset + field_ids_size > data_size) {
+    set_integer(1, module, "invalid_dex");
+    return;
+  }
+
+  FIELD_ID_ITEM *field_ids = yr_malloc(field_ids_size);
+  memcpy(field_ids, data + offset, field_ids_size);
+
+  field_ids_size = *dex_header->field_ids_size * sizeof(FIELD_ID_ITEM);
+  int i = 0;
+  int p = 0;
+  for (i = 0, p = 0; p < field_ids_size; i += 1, p += sizeof(FIELD_ID_ITEM)) {
+    uint16_t class_idx = *field_ids[i].class_idx;
```

Comment:
You should always construct the integer values by applying or operations on the bytes directly. You can ifdef assuming a host and target endian, but compilers should be smart enough to optimize those cases if you use some generic code for this.

You can have a look at the r_endian.h we use in radare2, to get an idea about the primitives you'll need and how to use them